File: Tasks/112651_Shortest route_Кратчайший маршрут/main.py
import math


def dijkstra(graph, start):
    min_lengths_to_nodes = {i + 1: math.inf for i in range(len(graph))}
    min_lengths_to_nodes[start] = 0

    shortest_path = {start: -1}

    visited_nodes = []
    current_node = start

    while len(graph) > len(visited_nodes) and current_node is not None:
        for neighbor_node in graph[current_node]:
            length_to_neighbor = min_lengths_to_nodes[current_node] + graph[current_node][neighbor_node]

            if min_lengths_to_nodes[neighbor_node] > length_to_neighbor:
                min_lengths_to_nodes[neighbor_node] = length_to_neighbor
                shortest_path[neighbor_node] = current_node

        visited_nodes.append(current_node)

        min_length = math.inf
        next_node = None
        for node in min_lengths_to_nodes:
            if min_lengths_to_nodes[node] < min_length and node not in visited_nodes:
                min_length = min_lengths_to_nodes[node]
                next_node = node

        current_node = next_node
    return min_lengths_to_nodes, shortest_path


def input_data():
    n = int(input())
    edges = [[int(i) for i in input().split()] for j in range(n)]
    start, end = [int(i) for i in input().split()]

    graph = {}
    for node in range(n):
        node_edges = {}
        for index, edge in enumerate(edges[node]):
            if edge == 0:
                continue
            node_edges[index + 1] = edge
        graph[node + 1] = node_edges

    return start, end, graph


def print_result(end, min_lengths_to_nodes, shortest_path):
    if min_lengths_to_nodes[end] == math.inf:
        print(0)
        return

    print(min_lengths_to_nodes[end])

    s = []
    s.insert(0, end)
    while shortest_path[end] != -1:
        end = shortest_path[end]
        s.insert(0, end)
    print(*s)


def main():
    start, end, graph = input_data()
    min_lengths_to_nodes, shortest_path = dijkstra(graph, start)
    print_result(end, min_lengths_to_nodes, shortest_path)


if __name__ == '__main__':
    main()

File: Tasks/112651_Shortest route_Кратчайший маршрут/main.cpp
#include <iostream>
#include <map>
#include <vector>
#include <climits>
#include<algorithm>
using namespace std;


pair<map<int, int>, map<int,int>> dijkstra(map <int, map<int, int>> graph, int start) {
    map <int, int> min_lengths_to_nodes;
    for (int i = 0; i < graph.size(); i++) {
        min_lengths_to_nodes[i + 1] = INT_MAX;
    }

    min_lengths_to_nodes[start] = 0;
    map <int, int> shortest_path;
    shortest_path[start] = -1;

    vector<int> visited_nodes;
    int current_node = start;
    int length_to_neighbor;

    while (graph.size() > visited_nodes.size() && current_node != NULL) {
        for (auto neighbor_node : graph[current_node]) {
            length_to_neighbor = min_lengths_to_nodes[current_node] + graph[current_node][neighbor_node.first];
            if (min_lengths_to_nodes[neighbor_node.first] > length_to_neighbor) {
                min_lengths_to_nodes[neighbor_node.first] = length_to_neighbor;
                shortest_path[neighbor_node.first] = current_node;
            }
        }
        visited_nodes.push_back(current_node);
        int min_length = INT_MAX;
        int next_node = NULL;
        for (auto node : min_lengths_to_nodes) {
            if (min_lengths_to_nodes[node.first] < min_length && find(visited_nodes.begin(), visited_nodes.end(), node.first) == visited_nodes.end())
            {
                min_length = min_lengths_to_nodes[node.first];
                next_node = node.first;
            }
        }

        current_node = next_node;
    }
    return pair<map<int, int>, map<int, int>>( min_lengths_to_nodes, shortest_path);
}

int main() {
    //input
    int n, start, end;
    cin >> n;
    vector < vector <int> > edges(n, vector <int>(n)); 

    for (int i = 0; i < n; i++)     
        for (int j = 0; j < n; j++) 
        {
            cin >> edges[i][j]; 
        }
    cin >> start >> end;
    map <int, map<int, int>> graph;
    for (int node = 0; node < n; node++) {
        map <int, int> node_edges;
        for (int index = 0; index < edges[node].size(); index++) {
            if (edges[node][index] == 0)
                continue;
            node_edges[index + 1] = edges[node][index];
        }
        graph[node + 1] = node_edges;
    }


    //print
    pair<map<int, int>, map<int, int>> para = dijkstra(graph, start);
    map <int, int> min_lengths_to_nodes = para.first;
    map <int, int> shortest_path = para.second;
    if (min_lengths_to_nodes[end] == INT_MAX) {
        cout << "0";
    }
    else {
        cout << min_lengths_to_nodes[end] << endl;
        vector<int> s;
        s.insert(s.begin(), end);
        while (shortest_path[end] != -1) {
            end = shortest_path[end];
            s.insert(s.begin(), end);
            
        }
        for (int i = 0; i < s.size(); i++) {
            cout << s[i] << " ";
    }
    }
    system("pause");
    return 0;
}